### PR TITLE
Major refactoring, implementing LBP

### PR DESCRIPTION
**Pull request by rvollgraf**
Implementing loopy belief propagation in factor graph model

### CODE REVIEW

File: code/experiments/exp_5.1/exp51_lbp.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+import argparse
+
+from matplotlib.pyplot import *
+from matplotlib import cycler
+
+import numpy as np
+import random
+
+
+def init_contacts(S, T, qIbar=20.0, R0: Union[float, np.array] = 2.5, p1=0.01, decay=0.1,
```
Comment (rherbrich74) [inline code/experiments/exp_5.1/exp51_lbp.py:20]: Why not compute **qIBar** once when **qI** gets defined and use it as an argument on each calling site?

File: code/experiments/exp_5.1/exp51_lbp.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    l = np.arange(S)
+    l0 = l[:,np.newaxis]
+    l1 = l[np.newaxis,:]
+
+    mask = (l[:,np.newaxis] > l[np.newaxis,:])
+    idx = list(zip(*np.where(mask)))
+
+    if H is not None:
+        maskb = mask * (l0 - l1 < l0 % H + 1)
+        maska = mask * (~maskb)
```
Comment (rherbrich74) [inline code/experiments/exp_5.1/exp51_lbp.py:44]: Can you comment these lines more - the reader (speak: I) struggle to understand how these work.

File: code/test_gibbs_sampler.py [hunk 3]
```diff
@@ -32,13 +32,15 @@ def nb_iter(n,p):
     T = 30
     S = 3
 
-    alpha = 0.001
-    beta = 0.01
+    alpha = 0.01
+    beta = 0.1
     p0 = 0.02
-    p1 = 0.99
+    p1 = 0.8
```
Comment (rherbrich74) [inline code/test_gibbs_sampler.py:38]: Why did you increase these?

File: code/test_gibbs_sampler.py [hunk 4]
```diff
@@ -113,23 +115,20 @@ def nb_iter(n,p):
 ### sample with the Gibbs sampler
 
 Nsamp = 10000
-crisp = PopulationInfectionStatus( S, T, contacts, tests,qE,qI, alpha, beta, p0, p1, False)
+crisp = GibbsPIS( S, T, contacts, tests,qE,qI, alpha, beta, p0, p1, False)
 
 t = time.time()
-Z = crisp.gibbs_sample(Nsamp)
+#Z = crisp.gibbs_sample(Nsamp)
```
Comment (rherbrich74) [inline code/test_gibbs_sampler.py:121]: I suggest to remove this line.

File: code/test_gibbs_sampler.py [hunk 5]
```diff
@@ -113,23 +115,20 @@ def nb_iter(n,p):
...
+for u,infs in enumerate(p[:,-1]):
     print("{}: {}".format(u,infs))
 
-p = (np.cumsum(Z,axis=2)[:,np.newaxis,:,:] > np.arange(T).reshape(1,-1, 1, 1)).mean(axis=0)
-p = np.diff(p,axis=2, prepend=0)
-p = np.concatenate( [p, 1.0-p.sum(2,keepdims=True)], axis=2)
-
 figure(figsize=(6,8))
 for i in range( min(5,S)):
     subplot( min(5,S),1,i+1)
```
Comment (rherbrich74) [inline code/test_gibbs_sampler.py:130]: I suggest to remove the additional spaces before **min**

File: code/test_gibbs_sampler.py [hunk 6]
```diff
@@ -172,22 +168,78 @@ def nb_iter(n,p):
 ### and parentCounts are created correctly, shouls look like the first figure
 
 t = time.time()
-Z = crisp_fwd.gibbs_sample(Nsamp)
+p = crisp_fwd.get_marginals(Nsamp)
 print("generated {} Gibbs samples in {:.3}s".format(Nsamp, time.time()-t))
 print("infection stati:")
-for u,infs in enumerate( crisp_fwd.get_infection_status(Nsamp)):
+for u,infs in enumerate( p[:,-1]):# crisp_fwd.get_infection_status(Nsamp)):
```
Comment (rherbrich74) [inline code/test_gibbs_sampler.py:174]: I suggest to remove the comment - it relates to a non-available call-syntax

File: code/test_gibbs_sampler.py [hunk 7]
```diff
@@ -172,22 +168,78 @@ def nb_iter(n,p):
...
     subplot( min(5,S),1,i+1)
-    plot(p[:,i],'.-')
+    plot(p[i],'.-')
     grid(True)
 suptitle("Gibbs sampling in iteratively built up CRISP")
 
 
+##########################################################################################
+### inference with the LBP model
+crisp = LBPPIS( S, T, contacts, tests,qE,qI, alpha, beta, p0, p1, False)
```
Comment (rherbrich74) [inline code/test_gibbs_sampler.py:187]: I suggest to remove the leading space at **S**

File: code/test_gibbs_sampler.py [hunk 8]
```diff
@@ -172,22 +168,78 @@ def nb_iter(n,p):
...
+
+t = time.time()
+p = crisp.get_marginals()
+print("LBP inference in {:.3}s".format(time.time()-t))
+print("infection stati:")
+for u,infs in enumerate(p[:,-1]):
+    print("{}: {}".format(u,infs))
+
+figure(figsize=(6,8))
+for i in range( min(5,S)):
```
Comment (rherbrich74) [inline code/test_gibbs_sampler.py:198]: I suggest to remove the leading space before **min**

File: code/test_gibbs_sampler.py [hunk 9]
```diff
@@ -172,22 +168,78 @@ def nb_iter(n,p):
...
+    subplot( min(5,S),1,i+1)
```
Comment (rherbrich74) [inline code/test_gibbs_sampler.py:199]: I suggest to remote the leading space before **min**

File: code/test_gibbs_sampler.py [hunk 10]
```diff
@@ -172,22 +168,78 @@ def nb_iter(n,p):
...
```
Comment (rherbrich74) [inline code/test_gibbs_sampler.py:219]: I suggest to remove the leading spaces.

File: code/test_gibbs_sampler.py [hunk 11]
```diff
@@ -172,22 +168,78 @@ def nb_iter(n,p):
...
+    plot(p[i],'.-')
+    grid(True)
+suptitle("Inference with Baum-Welch message passing (5 cycles)")
+
+##########################################################################################
+### inference with the LBP model
+crisp.reset()
+crisp.propagate(1,"forward")
+
```
Comment (rherbrich74) [inline code/test_gibbs_sampler.py:228]: I suggest to remove the additional empty space

File: code/test_gibbs_sampler.py [hunk 12]
```diff
@@ -172,22 +168,78 @@ def nb_iter(n,p):
...
+t = time.time()
+p = crisp.get_marginals()
+print("LBP inference in {:.3}s".format(time.time()-t))
+print("infection stati:")
+for u,infs in enumerate(p[:,-1]):
+    print("{}: {}".format(u,infs))
+
+figure(figsize=(6,8))
+for i in range( min(5,S)):
+    subplot( min(5,S),1,i+1)
```
Comment (rherbrich74) [inline code/test_gibbs_sampler.py:239]: I suggest to remove the leading spaces before **min**

File: code/test_gibbs_sampler.py [hunk 13]
```diff
@@ -172,22 +168,78 @@ def nb_iter(n,p):
...
+    plot(p[i],'.-')
+    grid(True)
+suptitle("Inference with forward message passing (1 cycle)")
+
+
```
Comment (rherbrich74) [inline code/test_gibbs_sampler.py:244]: I suggest to remove the empty lines.

File: code/test_gibbs_sampler.py [hunk 14]
```diff
@@ -138,31 +137,28 @@ def nb_iter(n,p):
 ### buildup CRIP many times with repeated calls to advance, thereby forward sampling
```
Comment (rherbrich74) [inline code/test_gibbs_sampler.py:137]: CRIP ==> CRISP

File: code/CMakeLists.txt [hunk 15]
```diff
@@ -14,4 +14,6 @@ FIND_PACKAGE(PythonInterp 3)
 
 project(crisp CXX)
 add_subdirectory(pybind11)
-pybind11_add_module(crisp crisp.cpp)
+pybind11_add_module(crisp python.cpp lbp.cpp gibbs.cpp distribution.cpp crisp.cpp node.cpp factor.cpp seir.cpp)
+
+add_executable (testit main.cpp node.cpp factor.cpp seir.cpp distribution.cpp)
```
[the remark this answers is not in the document]
Comment (rvollgraf) [reply code/CMakeLists.txt:19]: done

File: code/crisp.hpp [hunk 16]
```diff
@@ -0,0 +1,163 @@
+#ifndef _POPULATION_HPP
+#define _POPULATION_HPP
```
[the remark this answers is not in the document]
Comment (rvollgraf) [reply code/crisp.hpp:2]: makes sense ;)

File: code/crisp.hpp [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+
+#include <vector>
+#include <iostream>
+# include <random>
```
[the remark this answers is not in the document]
Comment (rvollgraf) [reply code/crisp.hpp:6]: done

File: code/crisp.hpp [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+    private:
+        int         _individual;
+        int         _time;
+        TestOutcome _outcome;
+    public:
+        Outcome(int u, int t, TestOutcome o) :
+            _individual(u),
+            _time(t),
+            _outcome(o) {}
+        Outcome(const OutcomeTuple &o): Outcome(get<0>(o), get<1>(o), get<2>(o)==0?Negative:Positive) {}
```
[the remark this answers is not in the document]
Comment (rvollgraf) [reply code/crisp.hpp:33]: thanks

File: code/crisp.hpp [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+        // Random number generator
+        random_device _rd;
+        mt19937 _gen;
+
+        // Contacts data
+        vector<vector<vector<tuple<int,int>>>> _contacts;
+        vector<tuple<int,int>> _empty;
+        inline vector<tuple<int,int>>& _futureContact(int u, int t) { return t>=0 && t<_noTimeSteps ? _contacts[u][t] : _empty ; }
+        inline vector<tuple<int,int>>& _pastContact(int u, int t) { return t>=1 && t<=_noTimeSteps ? _contacts[u][t-1] : _empty ; }
+
```
[the remark this answers is not in the document]
Comment (rvollgraf) [reply code/crisp.hpp:98]: seems we don't need them -> removed

File: code/distribution.hpp [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+        // Returns the log-probability mass function at value k
+        double getLogP(int k) const {
+            return((k<0)?(-INFINITY):((k)*log(1.0-_p0)+log(_p0)));
+        }
+
+        // Returns the log probability mass function of the truncated distribution at value k
+        double getLogPTail(int k) const {
+            return((k<0)?0:((k)*log(1.0-_p0)));
+        }
+};
```
[the remark this answers is not in the document]
Comment (rvollgraf) [reply code/distribution.hpp:63]: they are not used interchangeably, I would leave that as it is for now

File: code/experiments/exp_5.1/exp51_lbp.py (hunk 1 above)
Comment (rvollgraf) [reply code/experiments/exp_5.1/exp51_lbp.py:20]: I've left that unchanged from the previous version. We can address this in a later PR

File: code/seir.cpp [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+    return SEIRState(S);
+}
+
+
+bool SEIRState::operator == (const SEIRState &other) const {
+    if(_p==S && other._p==S) return true;
+    if(_p==R && other._p==R) return true;
+    if(_d != other._d) return false;
+    if(_p==E && other._p==E) return true;
+    if(_p==I && other._p==I) return true;
```
[the remark this answers is not in the document]
Comment (rvollgraf) [reply code/seir.cpp:41]: thanks

File: code/experiments/exp_5.1/exp51_lbp.py [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+    contacts_2 = init_contacts(S=S, T=T, R0=R0_2, seed=args.seed )
+    fig_2 = make_figure(contacts_2)
+    fig_2.gca().set_ylim(fig_1.gca().get_ylim())
+    fig_2.gca().axvline(x=60, color=[0.8, 0.8, 0.8], linestyle='--')
+    fig_2.gca().set_ylim(fig_2.gca().get_ylim())
+    title('suppression after 60 days')
+    print("ok")
+
+    print("creating figure 3 ... ", end="")
+    R0_3 = np.array([R0] * 60 + [0.5] * 60 + [2.5] * (T - 120))
```
[the remark this answers is not in the document]
Comment (rvollgraf) [reply code/experiments/exp_5.1/exp51_lbp.py:198]: yes

File: code/experiments/exp_5.1/exp51_lbp.py [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+             0.0589104177, 0.05476817903, 0.05015542853, 0.0453410888, 0.04053528452,
+             0.03589255717, 0.03151878504, 0.02747963753, 0.02380914891, 0.02051758911,
+             0.01759822872, 0.01503287457, 0.0127962154, 0.01085910889, 0.009190974483,
+             0.007761463001, 0.006541562648, 0.005504277076]
+
+    qE = Distribution([q/sum(qEVec) for q in qEVec])
+    qI = Distribution([q/sum(qIVec) for q in qIVec])
+
+
+    def make_figure( contacts, t_branch=None, contacts_branch=None):
```
[the remark this answers is not in the document]
Comment (rvollgraf) [reply code/experiments/exp_5.1/exp51_lbp.py:120]: done

File: code/experiments/exp_5.1/exp51_lbp.py (hunk 2 above)
Comment (rvollgraf) [reply code/experiments/exp_5.1/exp51_lbp.py:44]: done

File: code/test_gibbs_sampler.py (hunk 3 above)
Comment (rvollgraf) [reply code/test_gibbs_sampler.py:38]: leftover from experimenting, doesn't make a big difference, reverted

File: code/experiments/exp_5.1/exp51_lbp.py [hunk 24]
```diff
@@ -0,0 +1,212 @@
...
+        return fig
+
+
+    print("creating figure 0 ... ")
+    contacts = init_contacts(S=S, T=T, R0=R0, p1=p1, seed=args.seed)
+    fig_0 = make_figure(contacts)
+    title('no mitigation')
+    print("ok")
+    
+    print("creating figure 4 ... ", end="")
```
[the remark this answers is not in the document]
Comment (rvollgraf) [reply code/experiments/exp_5.1/exp51_lbp.py:171]: This is from the Gibbs version of the experiment, were fig. 4 indeed was following from fig. 0. I would leave it here as it is. Both files could be refactored together at a later time.

**Pull request merged**